File: go_algorithms/algorithms/RowByRowParallel.go
package algorithms

import (
	"fmt"
	"sync"
	"time"
)

func RowByRowParallel(B, C [][]int) ([][]int, int64) {
	fmt.Println("Método: RowByRowParallel")

	startTime := time.Now().UnixNano()

	size := len(B)
	A := make([][]int, size)
	for i := range A {
		A[i] = make([]int, size)
	}

	bsize := size / 2
	var wg sync.WaitGroup

	for i1 := 0; i1 < size; i1 += bsize {
		for j1 := 0; j1 < size; j1 += bsize {
			for k1 := 0; k1 < size; k1 += bsize {
				wg.Add(1)
				go func(i1, j1, k1 int) {
					defer wg.Done()
					for i := i1; i < i1+bsize && i < size; i++ {
						for j := j1; j < j1+bsize && j < size; j++ {
							for k := k1; k < k1+bsize && k < size; k++ {
								A[i][k] += B[i][j] * C[j][k]
							}
						}
					}
				}(i1, j1, k1)
			}
		}
	}

	wg.Wait()

	endTime := time.Now().UnixNano()
	executionTime := endTime - startTime

	return A, executionTime
}

File: python_algorithms/algorithms/RowByRowParallel.py
import numpy as np
import threading
import time

def Multiply(B, C):
    print("Método: RowByRowParallel")

    start_time = time.time_ns()

    size = len(B)
    A = np.zeros((size, size), dtype=np.int64)

    bsize = size // 2
    threads = []

    for i1 in range(0, size, bsize):
        for j1 in range(0, size, bsize):
            for k1 in range(0, size, bsize):
                thread = threading.Thread(target=worker, args=(i1, j1, k1, A, B, C, bsize, size))
                thread.start()
                threads.append(thread)

    # Esperar a que todos los hilos terminen
    for thread in threads:
        thread.join()

    end_time = time.time_ns()
    execution_time = end_time - start_time

    return A, execution_time

def worker(i1, j1, k1, A, B, C, bsize, size):
    for i in range(i1, min(i1 + bsize, size)):
        for j in range(j1, min(j1 + bsize, size)):
            for k in range(k1, min(k1 + bsize, size)):
                A[i][k] += B[i][j] * C[j][k]
